### DISCUSSION TITLE
Usage of stoichiometric matrix and kinetic information

### DISCUSSION THREAD

**Discussion by formersbach** [biomass-dev/biomass#175] [Ideas]
I opened this discussion to talk about potential uses of the stoichiometric matrix and kinetic information.
I would like to add every possible usage as a separate comment thread in this discussion.

**Comment by formersbach** [comment]
The stoichiometric matrix can be used to detect possible combinations of fluxes that can lead to steady states. 
Column vectors of the kernel of the stoichiometric matrix and their linear combinations are the only possible fluxes for steady state.
Therefore by solving
$\frac{dS}{dt} = N * K = 0$ with $S$ being species concentrations, $N$ being the stoichiometric matrix, and $K$ being the Kernel we can gain insights into the steady state. 
For example, by summing up the rows of K we can find out if there are reactions that need a flux of 0 for a valid steady state.
This can easily be solved using sympy:

```python
from biomass import Text2Model
import sympy
model = Text2Model("test.txt")
model.convert(overwrite=True)
 # transform to sympy matrix
stoichiometry = sympy.Matrix(model.stoichiometry)
# compute nullspace/kernel, returns list of matrix, transform to sympy matrix
steadystates = sympy.Matrix([stoichiometry.nullspace()]) 
possible_states = steadystates * sympy.ones(steadystates.shape[1], 1) # calculate row sum
```
For the toy model:

>      A <--> B
>      A is synthesized

the code above returns [1, 1, 0], indicating that A --> B and B --> A need to have equal flux, while the synthesis of A needs to be 0. 
If the Kernel only contains the 0 vector, no steady state is possible. 

I think this would be very useful information in the initial model generation.

**Reply by himoto** [reply]
Looks good!

**Comment by formersbach** [comment]
I would like to change how graphs are drawn.
Currently the ode.py/reaction_network.py files are parsed. 
With access to kinetic information this can be simplified by using reactants/products/modifiers.
For this we need to define how edges are drawn.
My initial suggestion would be [Reactants, Modifiers] --> [Products].

**Reply by himoto** [reply]
I completely agree with you. Let's do this in another pull request!
As we discussed, I would like to move graph visualization method to construction/.
Can I close #174 without merging? I would like to create a TODO list for the next updates since there will be a lot to change.

**Reply by formersbach** [reply]
I already closed it! And the corresponding issue as well.
I'll open a pull request for the changes to graphing.

**Comment by himoto** [comment]
**TODO:**
- [x] Transfer graphing from BioMass to Text2Model (in progress #177)
- [x] Build Oppelt et al's NF-kB model via Text2Model
- [x] Update [tutorial 2](https://biomass-core.readthedocs.io/en/latest/tutorial/nfkb.html)

**Reply by himoto** [reply]
@formersbach 
I am planning to release v0.10.0 when these TODO are completed.
